File: Crio/crio_programming_interview_problems-master/NetworkDelayTime/NetworkDelayTime.py

# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
from queue import PriorityQueue
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def networkDelayTime(n,edges,source):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    
    graph=[[] for _ in range(n+1)] 
    for i in range(len(edges)):
        a=edges[i][0]
        b=edges[i][1]
        c= edges[i][2]
        graph[a].append([b,c])
    q = PriorityQueue()

    distance=[2**60-1]*(n+1)
    distance[source]=0
    for b,c in graph[source]:
        q.put((c,b))
        distance[b]=c

    while not q.empty():
        dis,node=q.get()
        distance[node] = min(distance[node],dis)
        for b,c in graph[node]:
            if distance[b] > distance[node]+c:
                distance[b]= distance[node]+c
                q.put((distance[b],b))

    cost=-1
    for i in range(1,n+1):
        cost= max(cost,distance[i])
    if cost == (2**60-1):
        cost=-1
    return cost
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    n,e=map(int,input().split())
    edges=[]
    for i in range(e):
        a,b,c= map(int,input().split())
        edges.append([a,b,c])
    s=int(input())
    ans = networkDelayTime(n,edges,s)
    print(ans)
if __name__ == '__main__':
    main()

File: Crio/crio_programming_interview_problems-master/NetworkDelayTime/NetworkDelayTime.cpp
#include<bits/stdc++.h>
using namespace std;

int networkDelayTime(int n, vector<vector<int>>& times, int K){
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int nodes = n;
    int source = K;
    vector<vector<int>>& edges=times;
    vector<vector<pair<int,int>>>adj(nodes+1);
    for (auto &&edge : edges)
    {
        adj[edge[0]].push_back({edge[1],edge[2]});
    }
    vector<int>dist(nodes+1,INT_MAX);
    dist[source]=0;
    priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>>q;
    for (auto &&edge : adj[source])
    {
        int time = edge.second;
        int child = edge.first;
        dist[child] = time;
        q.push({time,child});
    }
    while (!q.empty())
    {
        pair<int,int> x = q.top();
        q.pop();
        int time = x.first;
        int node = x.second;
        for (auto &&edge : adj[node])
        {
            int child = edge.first;
            int d = edge.second;
            if(dist[child]>d+time){
                dist[child] = d+time;
                q.push({dist[child],child});
            }
        }
    }
    int ans = *max_element(1+dist.begin(),dist.end());
    if(ans == INT_MAX) ans=-1;
    return ans;
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

int main(){

    int nodes, source, e;
    cin>>nodes>>e;
    vector<vector<int>>edges(e);
    for (int i = 0; i < e; i++)
    {
        int u,v,t;
        cin>>u>>v>>t;
        edges[i].push_back(u);
        edges[i].push_back(v);
        edges[i].push_back(t);
    }
    cin>>source;
    int ans = networkDelayTime(nodes,edges,source);
    cout<<ans;
}
